Add lint to improve floating-point expressions

krishna-veerareddy opened the pull request:
Looks for floating-point expressions that can be expressed using built-in methods to improve accuracy, performance and/or succinctness.

changelog: Add lint `floating_point_improvements`.

Fixes #4726 
Partly addresses [#2040](https://github.com/rust-lang/rust-clippy/issues/2040)

Currently linted expressions:

| Expression | Suggestion |
|---------------------------------|------------|
| x.log(2.0) | x.log2() |
| x.log(10.0) | x.log10() |
| x.log(std::f32::consts::E) | x.ln() |
| (1 + x).ln() | x.ln_1p() |
| (2.0).powf(x) | x.exp2() |
| (std::f32::consts::E).powf(x) | x.exp() |
| x.powf(1/2) | x.sqrt() |
| x.powf(1/3) | x.cbrt() |
| x.powf(y), where y is whole | x.powi(y) |
| x.exp() - 1 | x.exp_m1() |
|x * y + z|x.mul_add(y, z)|

krishna-veerareddy commented:
@clarfon Hey I wanted to get the ball rolling on this [issue](https://github.com/rust-lang/rust-clippy/issues/2040) but I have a few questions regarding how we would separate the lints. For example, should I create separate lints for `log(2|10|e)` and `log_1p` functions or should I group all the log functions under a single lint or should I group all the suggestions under a single lint?

The direction I was going in was creating separate lints for each of your suggestions but I am not sure if that's the right approach.

clarfon commented:
Wouldn't this be the same as one of the known constant lints? Don't see why this would be lumped under a different lint.

We could maybe even lint common trig function calls which return values of already provided constants, e.g. `sin(PI) == 0`. But maybe calling it a "computed known constant" rather than a literal without a name is better, and splitting that from the approx literal lint.

krishna-veerareddy commented:
No, this would lint calls such as `(1f32).log(2.0)` and suggest usage of `(1f32).log2()` instead. It does the same for bases `10` and `e`. `known_log_bases` was probably a poor choice for the lint name.

clarfon commented:
Oh! I did that thing I absolutely shouldn't do which is to assume something without reading the whole thing. -___-

Honestly, I feel like one lint should cover all cases for things like this. Things that have explicit specialised versions should go in one issue.

Things which don't have explicit functions but are more accurate should go in a separate lint, e.g. Herbie's home page example of `sqrt(x+1) - sqrt(x)` being better off as `1/(sqrt(x+1) + sqrt(x))`.

clarfon commented:
Another potential thing to consider is whether you're optimising for speed or accuracy. I'd maybe separate lints based upon that, e.g. `hypot` is always more accurate but may be slower.

krishna-veerareddy commented:
@clarfon Hey I created two lints: one for better accuracy and another for performance. The expression currently linted are:
```
x.log(2) => x.log2() // for better accuracy
x.log(10) => x.log10() // for better accuracy
x.log(E) => x.ln() // for better accuracy
(1 + x).ln() => x.ln_1p() // for better accuracy
(2.0).powf(x) => x.exp2() // for better performance
E.powf(x) => x.exp() // for better performance
x.powf(1 / 2) => x.sqrt() // for better performance
x.powf(1 / 3) => x.cbrt() // for better performance
x.exp() - 1 => x.exp_m1() // for better accuracy
```
I will add lints for some of the other suggestions in your issue in a separate PR. Would love to know your thoughts/suggestions.

clarfon commented:
Honestly, it may be worth a "always better" lint for some. For example, the log lints are pretty much the same or better for performance and better for accuracy. Ditto for the log plus one and exp minus one lints. However, changing to hypot or fused multiply-add might reduce performance, and thus is better for accuracy only.

krishna-veerareddy commented:
Yeah makes sense. Any suggestion for the lint name?

llogiq commented:
How about `floating_point_perf` and `floating_point_accuracy`?

krishna-veerareddy commented:
@llogiq Any suggestion for lints that are better for both performance and accuracy? Should I use `floating_point_perf_and_accuracy` or is that too verbose.

krishna-veerareddy commented:
Having a `floating_point_improvements` lint for better accuracy and performance sounds good to me. Should this lint also cover cases where there is no clear accuracy/performance improvement but makes the code more idiomatic or should that be a separate lint? Here are a couple examples:
```
x * PI / 180 => x.to_radians()
x * 180 / PI => x.to_degrees()
x.logN() / y.logN() => x.logbase(y)
```

llogiq commented:
I'd put everything that sacrifices accuracy under `floating_point_performance`, everything that sacrifices performance under `floating_point_accuracy` and everything else under `floating_point_improvements`.

krishna-veerareddy commented:
I didn't have to add the `floating_point_accuracy` and `floating_point_performance` lints yet but please let me know if any of the methods in this PR would sacrifice either performance or accuracy and Ill add that method under the relevant lint.

bors commented:
:umbrella: The latest upstream changes (presumably #4930) made this pull request unmergeable. Please resolve the merge conflicts.

flip1995 commented:
I agree with the grouping of the lints, but I don't like the names. E.g., `#[allow(floating_point_improvements)]` is kind of misleading and against our lint naming convention (cc #2845).

My suggestions:

- accuracy: `inaccurate_floating_points`
- perf: `bad_performing_floating_points`
- general: `bad_floating_points`

I'm not happy with the word `bad` though. But writing `allow(lint_name)` should describe (in an English "sentence") what is allowed there:

```rust
#[allow(clippy::inaccurate_floating_point)]
let y = x.log(2);
```

So it is obvious, that `x.log(2)` is inaccurate, but acceptable there, whereas `allow(floating_point_accuracy)` kind of suggests the opposite.

krishna-veerareddy commented:
Makes sense. How about:
general: `{inferior|substandard|ill_advised|misguided}_floating_points`
accuracy: `inaccurate_floating_points`
performance: `{slow|inefficient}_floating_points`

Also having a `floating_point_{operation|computation|arithmetic|expression}` suffix makes more sense to me than `floating_points` although it does make the lint name long.

flip1995 commented:
I like `inefficient_floating_points` as a lint name for the perf group. For the general case I guess `inferior` is the best word here.

As for the suffix, I personally like `operations` best.

clarfon commented:
Could always just do flops for floating point operations

krishna-veerareddy commented:
That is more concise but isn't [flops](https://en.wikipedia.org/wiki/FLOPS) supposed to be floating point operations per second?

llogiq commented:
Yes, {`inefficient`|`inaccurate`|`substandard`}`_flop` looks acceptable to me.

bors commented:
:umbrella: The latest upstream changes (presumably #4867) made this pull request unmergeable. Please resolve the merge conflicts.

clarfon commented:
IMHO, flops is FLoating Point Operations and flop/s is FLoating Point Operations per Second. However, some people use flops as FLoating Point Operations Per Second.

I guess that it's a matter of taste. I don't mind using flop or flops for the name.

krishna-veerareddy commented:
So to summarize:
- General: `substandard_flops`
- Efficiency over accuracy: `inefficient_flops`
- Accuracy over efficiency: `inaccurate_flops`

If we all unanimously agree to this I can go ahead and make these changes. Thanks!

clarfon commented:
I feel like `inaccurate` should be reworded to `imprecise`. `inaccurate` to me just feels like you're doing the wrong thing, rather than specifically having fewer accurate bits.

I feel like `suboptimal` could be worded as `always_suboptimal` or maybe just `always_worse`. Because optimal really depends on whether you value accuracy or speed, and some people might get the wrong impression. Adding an `always` implies that it doesn't matter what kind of optimal you're going for, and that it's always suboptimal. If that can be worded in one word but a different word, I'm cool with that.

krishna-veerareddy commented:
`imprecise` makes sense to me. When you say `suboptimal` are you referring to `substandard`? `suboptimal` hasn't come up in the discussion thus far.

clarfon commented:
Basically, I think it's important to somehow stress that they're both inefficient and inaccurate. But I don't want to put both those words in there because then the ordering of the two is arbitrary.

krishna-veerareddy commented:
Can't we stress that the expression is both inefficient and inaccurate in the lint help and have the lint name be something broad such as `inferior_flops` or `substandard_flops`?

The other issue is that currently the the general lint also includes style improvements(with no benefit to accuracy or performance) along with efficient/accuracy improvements so not sure what word would encompass these three things. Maybe we should start thinking about splitting the style improvements into its own lint and then come up with a word for the lint name that improves accuracy/performance.

flip1995 commented:
I would be good with `suboptimal` (I prefer this to `substandard`), `inefficient` and `imprecise`. 

Regarding style floats: I'm not sure if we should handle those separately. What would be example for style floats?

krishna-veerareddy commented:
Something like:  `x * (π / 180.0)` => `x.to_radians()`.

bors commented:
:umbrella: The latest upstream changes (presumably #4966) made this pull request unmergeable. Please resolve the merge conflicts.

clarfon commented:
(Comments on the choice suboptimal are fine; by no means feel like you're obligated to go with my ideas. :p)

I will add that the style-based ones can be problematic due to the fact that they don't affect precision or accuracy but may be different slightly due to the fact that FLOPs aren't associative. So, I agree that the first two could be in a "style" lint that isn't always used. The third is always better though, and is part of substandard.

flip1995 commented:
Well then we have to find a 4th name for the stylish flops 😄

krishna-veerareddy commented:
@flip1995 Hey can you take a look when you have some time. The lint is currently in nursery as there are some TODO's I need to fix before moving it out of there but I would like to get the initial changes merged.

flip1995 commented on clippy_lints/src/floating_point_arithmetic.rs:
```diff
@@ -272,6 +274,52 @@ fn check_expm1(cx: &LateContext<'_, '_>, expr: &Expr<'_>) {
+}
+
+// TODO: Fix rust-lang/rust-clippy#4735
+fn check_fma(cx: &LateContext<'_, '_>, expr: &Expr<'_>) {
```
NIT:
```suggestion
fn check_mul_add(cx: &LateContext<'_, '_>, expr: &Expr<'_>) {
```

Also can you rename the test files`*_fma`->`*_mul_add` for clarity.

flip1995 commented on clippy_lints/src/floating_point_arithmetic.rs:
```diff
@@ -0,0 +1,341 @@
+fn check_expm1(cx: &LateContext<'_, '_>, expr: &Expr<'_>) {
+    if_chain! {
+        if let ExprKind::Binary(op, ref lhs, ref rhs) = expr.kind;
+        if op.node == BinOpKind::Sub;
```
```suggestion
        if let ExprKind::Binary(BinOpKind::Sub, ref lhs, ref rhs) = expr.kind;
```

flip1995 commented on clippy_lints/src/floating_point_arithmetic.rs:
```diff
@@ -0,0 +1,341 @@
+fn check_ln1p(cx: &LateContext<'_, '_>, expr: &Expr<'_>, args: &[Expr<'_>]) {
+    if_chain! {
+        if let ExprKind::Binary(op, ref lhs, ref rhs) = &args[0].kind;
+        if op.node == BinOpKind::Add;
```
```suggestion
        if let ExprKind::Binary(BinOpKind::Add, ref lhs, ref rhs) = &args[0].kind;
```

flip1995 commented on tests/ui/floating_point_powf.rs:
```diff
@@ -0,0 +1,40 @@
+    let _ = x.powf(-2.0);
+    let _ = x.powf(16_777_215.0);
+    let _ = x.powf(-16_777_215.0);
+    let _ = x.powf(2.1);
```
```suggestion
    // Cases where the lint shouldn't be applied
    let _ = x.powf(2.1);
```

flip1995 commented on tests/ui/floating_point_powf.rs:
```diff
@@ -0,0 +1,40 @@
+    let _ = x.powf(-2.0);
+    let _ = x.powf(-2_147_483_648.0);
+    let _ = x.powf(2_147_483_647.0);
+    let _ = x.powf(2.1);
```
```suggestion
    // Cases where the lint shouldn't be applied
    let _ = x.powf(2.1);
```

flip1995 commented on clippy_lints/src/floating_point_arithmetic.rs:
```diff
@@ -0,0 +1,341 @@
+}
+
+// TODO: Lint expressions of the form `(x + y).ln()` where y > 1 and
+// suggest usage of `(x + (y - 1)).ln_1p()` instead
```
Has this better performance? If not, I'm not sure if we should suggest this, since it increases the complexity. I have the same concern with `expm1`

krishna-veerareddy replied:
I think it would be more accurate. I got this idea from the example documentation for `exp_m1` here: https://doc.rust-lang.org/std/primitive.f64.html#method.exp_m1 so I thought it would be something we could lint as well but I see your point. I can remove the TODO's for both `ln_1p` and `exp_m1`.

flip1995 replied:
You can keep the `TODO`s. We could still introduce a pedantic lint `suboptimal_floats_with_a_suggestion_that_makes_the_resulting_expression_more_complex` or another name.

krishna-veerareddy replied:
Sounds good. Thanks!

krishna-veerareddy commented:
@flip1995 Hey I changed the name of the lint to suboptimal_flops as we discussed above. The reason you don't see the other lints yet(`inefficient_flops` and `imprecise_flops`) were that all the checks above improve either accuracy or performance or both without loss of either. 

But, I looked at the implementation of these library functions and some of these are internally calling the `cmath` library which could go under the `imprecise_flops` category as the `cmath` implementation sacrifices performance for accuracy. Here are the methods that use `cmath` internally: `cbrt`, `ln_1p`, `exp_m1` and `hypot`(this one isn't linted yet). I can start making this change but would love to know your thoughts.

flip1995 commented:
>The reason you don't see the other lints yet [...] were that all the checks above improve either accuracy or performance or both without loss of either.

Ah ok, that's fair. As long as it is in nursery, it doesn't matter too much anyway.

> I can start making this change but would love to know your thoughts.

I think this is part of the split up, we should do, when we "stabilize" the lint(s)

> Also, I temporarily removed the lints to improve style as there were some bugs in implementation.

:+1: 

---

I'd say with the NITs fixed, I noted above, we can merge this, and everything else can be done in a later PR(s) :+1:

krishna-veerareddy commented:
@flip1995 Hey I created the `imprecise_flops` lint and moved some suggestions under that lint since I already had the changes. Also, fixed all the above comments. Please review when you have some time. Thanks!

flip1995 commented:
@bors r+

Thanks!

bors commented:
:pushpin: Commit ff0d44e45a27e688b5ee5447dc9175d5e72b37c9 has been approved by `flip1995`

<!-- @bors r=flip1995 ff0d44e45a27e688b5ee5447dc9175d5e72b37c9 -->
<!-- homu: {"type":"Approved","sha":"ff0d44e45a27e688b5ee5447dc9175d5e72b37c9","approver":"flip1995"} -->

bors commented:
:hourglass: Testing commit ff0d44e45a27e688b5ee5447dc9175d5e72b37c9 with merge 2734e4e1f7faf59b4b7c9db7070626b8b7fd2145...
<!-- homu: {"type":"BuildStarted","head_sha":"ff0d44e45a27e688b5ee5447dc9175d5e72b37c9","merge_sha":"2734e4e1f7faf59b4b7c9db7070626b8b7fd2145"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://help.github.com/en/actions), [checks-action_remark_test](https://help.github.com/en/actions), [checks-action_test](https://help.github.com/en/actions)
Approved by: flip1995
Pushing 2734e4e1f7faf59b4b7c9db7070626b8b7fd2145 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"flip1995","base_ref":"master","builders":{"checks-action_test":"https://help.github.com/en/actions","checks-action_dev_test":"https://help.github.com/en/actions","checks-action_remark_test":"https://help.github.com/en/actions"},"merge_sha":"2734e4e1f7faf59b4b7c9db7070626b8b7fd2145"} -->
